### PR TITLE
Signup/login

### PR DESCRIPTION
**Pull request by joelmackenz**
FE of signup and login pages complete, MongoDB Atlas connected

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -3,14 +3,23 @@ const express = require("express");
...
+const signupRouter = require("./routes/signup");
 
 const { json, urlencoded } = express;
 
 var app = express();
 
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+
+const connectionString = 'mongodb+srv://' + process.env.REACT_APP_MONGO_ID + ':' + process.env.REACT_APP_MONGO_PW + '@cluster0.q0hmz.mongodb.net/project?retryWrites=true&w=majority'
```
Comment (jessdesr) [inline server/app.js:20]: This can be moved fully to the environment file as (MONGO_DB_URI) instead of two separate for ID and password

File: client/src/pages/signin.css [hunk 2]
```diff
@@ -0,0 +1,100 @@
+.main {
```
Comment (sina-jamshidi) [inline client/src/pages/signin.css]: we're not going to use `.css` in our projects. Take a look at the Material UI [styles](https://material-ui.com/styles/basics/) page. We mainly want to use `makeStyles` and feed them into a `className` prop on each component

File: client/src/pages/signin.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
+import { Button, TextField, Paper, FormControl } from '@material-ui/core';
+import laptopPhoto from '../images/blue-shirt-at-laptop.png'
+import { Link } from 'react-router-dom';
+import './signin.css';
+
+export default class Signin extends React.Component {
+    render() {
+        return (
+            <div className="main">
```
Comment (sina-jamshidi) [inline client/src/pages/signin.js]: When using Material UI, try to use as many of its components as possible instead of generic HTML components.
`div`s can be [Grid](https://material-ui.com/components/grid/) and try to put all text in [Typography](https://material-ui.com/components/typography/). Whenever making a new component, just take a look at the list of components in Material UI and see if there's something you can use. Consistently using their components means a lot of styling will be built into the components already and it will make life easier

File: client/src/pages/signin.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
+import { Button, TextField, Paper, FormControl } from '@material-ui/core';
+import laptopPhoto from '../images/blue-shirt-at-laptop.png'
+import { Link } from 'react-router-dom';
+import './signin.css';
+
+export default class Signin extends React.Component {
```
Comment (sina-jamshidi) [inline client/src/pages/signin.js]: Try to use functional components with hooks.
[Official React docs](https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both) recommend that all new component be written with hooks, as that is the direction React is taking, and class based components will likely stay supported but outdated

File: client/src/pages/signup.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    const handleSubmit = (e) => {
+        e.preventDefault();
+        if (password === confirmPassword) {
+            axios.post('/signup', {
+                firstName: firstName,
+                lastName: lastName,
+                email: email,
+                password: password
+            })
+                .then(window.location = "/")
```
Comment (sina-jamshidi) [inline client/src/pages/signup.js]: We are using `react-router-dom` for routing in this project. Try to stay consistent with our routing methods to avoid problems later. You can use a conditional and the [Redirect](https://reactrouter.com/web/api/Redirect) component together in the `return` to redirect when a certain condition is met, in this case when the user has successfully logged in

File: client/src/pages/signup.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        } else {
+            alert("Passwords do not match")
```
Comment (sina-jamshidi) [inline client/src/pages/signup.js]: this can be discussed with Jess and your team about when you want to implement a better solution, but `alert` should definitely not be the long term solution here.

File: client/src/pages/signup.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+    return (
+        <div className="main">
+            <Paper className="image-container" variant="outlined">
+                <img className="laptop-image" src={laptopPhoto} alt="" />
+            </Paper>
+            <div className="right-screen">
+                <div className="to-sign-in">
+                    <p>Already have an account?</p>
+                    <Link to="/signin"><Button variant="outlined">Sign In</Button></Link>
```
Comment (sina-jamshidi) [inline client/src/pages/signup.js:165]: something I think you can do here is
```javascript
<Button to="/signin" component={Link} variant="outlined">Sign In</Button>
```

File: server/routes/signup.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+let express = require('express');
```
Comment (sina-jamshidi) [inline server/routes/signup.js:1]: are you supposed to handle all of auth in this PR? In that case the signin route is missing

File: server/routes/signup.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    res.send('Test response');
+  })
+
+  .post((req, res) => {
+    const newUser = new User(req.body);
+    newUser.save()
+      .then(() => {
+        res.statusCode = 200
+        res.setHeader("Content-Type", "text/html");
+        res.redirect('/')
```
Comment (sina-jamshidi) [inline server/routes/signup.js:18]: we are not handling routing from the server side, so we don't need this

File: server/routes/signup.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+const User = require('../models/user')
+
+router.route('/')
+  .get((req, res, next) => {
+    res.setHeader('Content-Type', 'application/json');
+    res.send('Test response');
+  })
+
+  .post((req, res) => {
```
Comment (sina-jamshidi) [inline server/routes/signup.js:12]: we should do more validations here before adding a user. Also remember to hash the password.
We also need some kind of token system for keeping a user logged in. Read the details of ticket #17 carefully.
If these are not supposed to be in this PR then try to keep the scope limited and keep these things for another PR

**Review by sina-jamshidi — CHANGES_REQUESTED** [review]

File: server/routes/signup.js (hunk 10 above)
Comment (joelmackenz) [reply server/routes/signup.js:12]: Oh, I only meant for this PR to be for the front-end and mongodb connection. I misunderstood what I was to push. I have not completed #17 yet.

File: client/src/pages/signup.js (hunk 7 above)
Comment (joelmackenz) [reply client/src/pages/signup.js:165]: This made a part of my useStyles (& Button), targeting the button as a child of another class, not work. I tried making it both & Button and & Link, but the Button would not be targeted. I don't feel too strongly either way, but if it's all the same, this saves one extra className.

File: server/app.js (hunk 1 above)
Comment (jessdesr) [reply server/app.js:20]: This is resolved in the BE branch

File: client/src/pages/signin.js [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        e.preventDefault();
+        axios.post('/api/signin', {
+            email: email,
+            password: password
+        }, {
+            withCredentials:true,
+            credentials:'include'
+        })
+            .then((response) => {
+                console.log(response);
```
Comment (sina-jamshidi) [inline client/src/pages/signin.js]: can remove the console.log

File: client/src/pages/signup.js [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+                password: false,
+                confirmPassword: false
+                // The request will only be verified/sent after Confirm Password has been entered correctly.
+            }), sendRequest(e));
+        }
+    }
+
+    const sendRequest = (e) => {
+        e.preventDefault()
+        for (var i in errors) {
```
Comment (sina-jamshidi) [inline client/src/pages/signup.js]: `let i` instead of `var i`

File: client/src/pages/signup.js [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+            if (errors[i] === true) {
+                console.log("Not all fields filled in")
+                var error;
```
Comment (sina-jamshidi) [inline client/src/pages/signup.js]: `const`

File: client/src/pages/signup.js [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+                return false;
+            }
+        }
+        // Only sends if all errors are false
+        if (!error) {
```
Comment (sina-jamshidi) [inline client/src/pages/signup.js:138]: just setting `const error` will still be undefined. So you should be setting `const error = true;`

**Review by sina-jamshidi — APPROVED** [review]
Couple of small notes to fix then good to go

File: client/src/pages/signup.js (hunk 14 above)
Comment (joelmackenz) [reply client/src/pages/signup.js:138]: When I set this to const error = true, the if-then statement that follows it cannot read its value. It seems to only read it if it is set to var. What can I do?

Comment (joelmackenz) [reply client/src/pages/signup.js:138]: Actually, I think I solved it. I changed it to a let, and moved the error back one step in the scope, so that the let and the if-then can access it.
